docs/Modeling-Artifacts.md

```diff
@@ -0,0 +1,367 @@
+
+## Graph Concepts
+
+A complex DAG may integrate artifacts, their referrers, and the Indexes that reference them. The following example demonstrates such a graph:
+
+```mermaid
+graph TD;
+
+I0["Index i0"]--manifests-->M0
+I0--manifests-->M1
+
+M2["Manifest m2"]--config-->Blob0
+M2--layers-->Blob5["Blob b5"]
+M2--subject-->M0
+
+M1["Manifest m1"]--config-->Blob3["Blob b3"]
+M1--layers-->Blob4["Blob b4"]
+
+M0["Manifest m0"]--config-->Blob0["Blob b0"]
+M0--layers-->Blob1["Blob b1"]
+M0--layers-->Blob2["Blob b2"]
+```
+
+For any node in the graph, the following definitions apply:
+
+- **Successor:** Any node that is pointed to by a given node. For instance:
+  - Blob `b0` is a successor of both `m0` and `m2`
+  - Manifest `m0` is a successor of `m2` and `i0`
+  
+- **Predecessor:** Any non-leaf node that directly points to a given node. For instance:
+  - Manifest `m0` is a predecessor of `b0`, `b1`, and `b2`
+  - Manifest `m2` is a predecessor of `b0`, `b5`, and `m0`
+  - Index `i0` is a predecessor of `m0` and `m1`
+
+These definitions apply to nodes of any type—manifests, indexes, or arbitrary blobs.
+
+However, the referrer relationship is different. A manifest (including [Image Manifest](https://github.com/opencontainers/image-spec/blob/v1.1.1/manifest.md) and [Index](https://github.com/opencontainers/image-spec/blob/v1.1.1/image-index.md)) with a `subject` field is considered a referrer of that subject manifest. According to [OCI image-spec v1.1.1](https://github.com/opencontainers/image-spec/blob/v1.1.1/manifest.md), both the `referrer` and the `subject` must be manifests.
+
+So, it is worth noting that:
+
+- `m0` is a `subject` of `m2`, and it is a successor of both `m2` and `i0`
+- `m2` is a referrer of `m0`, and it is a predecessor of `m0`, `b0`, and `b5`
+
+Defining functions `Predecessors()`, `Successors()`, and `Referrers()`, the example result would be:
+
+```
+Successors(m0) == [b0, b1, b2]
+Predecessors(m0) == [m2, i0]
+Referrers(m0) == [m2]
+
+Successors(m2) == [m0, b0, b5]
+Predessors(m2) == []
+Referrers(m2) == []
+
+Successors(b0) == []
+Predecessors(b0) == [m0, m2]
+Referrers(b0) == []
+```
+
+### Copy
+
+Given the root node of a Directed Acyclic Graph (DAG), the `Copy` function replicates the graph reachable from that root node from one Content-Addressable Storage (CAS) to another. This is achieved by recursively invoking the `Successors()` function to traverse and copy all descendant nodes in a certain order.
+
+Taking the graph above as an example:
+
+`Copy(m0)` copies the graph rooted by the node `m0`, including `m0` itself and all of its successors `b0`, `b1`, and `b2`.
+
+```mermaid
+graph TD;
+
+M0["Manifest m0"]--config-->Blob0["Blob b0"]
+M0--layers-->Blob1["Blob b1"]
+M0--layers-->Blob2["Blob b2"]
+```
+
+`Copy(m2)` copies the graph rooted by the node `m2`, including itself and all nodes reachable from it.
```

Comment:
How do you define reachable? Do you mean the nodes in the red cycle or blue? I assume it is blue.

![image](https://github.com/user-attachments/assets/e9e4f6fe-f8c4-4d3e-85c5-fc99514f8a86)

> including itself and all nodes reachable from it.

Reply:
The whole graph. Maybe I should make it clear that all nodes showed in the graph will be copied.